File: vol1/ch8/LeetCode855.py
from sortedcontainers import SortedList
class ExamRoom:
    def __init__(self, n: int):
        self.start,self.end=0,n-1
        self.tset=SortedList()

    def seat(self) -> int:
        if len(self.tset)==0:
            self.tset.add(0)
            return 0
        ans,pre=self.start,self.start
        maxdist=self.tset[0]-self.start
        for e in self.tset:
            if maxdist<(e-pre)//2:
                maxdist=(e-pre)//2
                ans=pre+maxdist
            pre=e
        if maxdist<self.end-pre:ans=self.end
        self.tset.add(ans)
        return ans

    def leave(self, p: int) -> None:
        self.tset.remove(p)

File: vol1/ch8/LeetCode855.cpp
class ExamRoom {
    set<int> tset;
    int start,end;
public:
    ExamRoom(int n) {
        start=0;
        end=n-1;
    }
    
    int seat() {
        int ans=start,pre=start;
        int maxdist=(tset.empty()?end:*tset.begin())-start;
        for(int e:tset) {
            if(maxdist<(e-pre)/2) {
                maxdist=(e-pre)/2;
                ans=pre+maxdist;
            }
            pre=e;
        }
        if(maxdist<end-pre)
            ans=end;
        tset.insert(ans);
        return ans;
    }
    
    void leave(int p) {
        tset.erase(p);
    }
};

